### PR TITLE
Add support for fibre channel

### PR DESCRIPTION
**Pull request by deusnefum**
Assuming there's no problems and this PR is accepted, I have a matching PR for node_exporter.

This resolves node_exporter feature request: prometheus/node_exporter#570

(I closed my astoundingly messed up previous PR)

### CODE REVIEW

**Comment by deusnefum** [comment]
I was going to whine about having add a test, but I ended up finding a bug so... point taken.

**Comment by pgier** [comment]
@discordianfish @SuperQ any concerns about merging this?

File: sysfs/class_fibrechannel.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+		if err != nil {
+			if os.IsNotExist(err) || err.Error() == "operation not supported" || err.Error() == "invalid argument" {
+				continue
+			}
+			return nil, fmt.Errorf("failed to read file %q: %v", name, err)
+		}
+
+		vp := util.NewValueParser(value)
+
+		// Below switch was automatically generated. Don't need everything in there yet, so the unwanted bits are commented out
```
Comment (SuperQ) [inline sysfs/class_fibrechannel.go:170]:
```suggestion
		// Below switch was automatically generated. Don't need everything in there yet, so the unwanted bits are commented out.
```

File: sysfs/class_fibrechannel.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+			return nil, err
+		}
+
+		fcc[host.Name] = *host
+	}
+
+	return fcc, nil
+}
+
+// Parse a single FC host
```
Comment (SuperQ) [inline sysfs/class_fibrechannel.go:86]:
```suggestion
// parseFibreChannelHost parses a single FC host.
```

File: sysfs/class_fibrechannel.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	FabricName       string               // /sys/class/fc_host/<Name>/fabric_name
+	DevLossTMO       string               // /sys/class/fc_host/<Name>/dev_loss_tmo
+	SupportedClasses string               // /sys/class/fc_host/<Name>/supported_classes
+	SupportedSpeeds  string               // /sys/class/fc_host/<Name>/supported_speeds
+	Counters         FibreChannelCounters // /sys/class/fc_host/<Name>/statistics/*
+}
+
+type FibreChannelClass map[string]FibreChannelHost
+
+// parses everything in /sys/class/fc_host
```
Comment (SuperQ) [inline sysfs/class_fibrechannel.go:64]:
```suggestion
// FibreChannelClass parses everything in /sys/class/fc_host.
```

File: sysfs/class_fibrechannel.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	counters, err := parseFibreChannelStatistics(path)
+	if err != nil {
+		return nil, err
+	}
+	host.Counters = *counters
+
+	return &host, nil
+}
+
+func parseFibreChannelStatistics(hostPath string) (*FibreChannelCounters, error) {
```
Comment (SuperQ) [inline sysfs/class_fibrechannel.go:145]:
```suggestion
// parseFibreChannelStatistics parses metrics from a single FC host.
func parseFibreChannelStatistics(hostPath string) (*FibreChannelCounters, error) {
```

**Comment by deusnefum** [comment]
That should do it. Sorry about that. Once this is accepted is their a guide or info on vendoring procfs into node_exporter? Is it -mod=vendor or something else?

File: sysfs/class_fibrechannel.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+	for _, f := range files {
+		if !f.Mode().IsRegular() || f.Name() == "reset_statistics" {
+			continue
+		}
+
+		name := filepath.Join(path, f.Name())
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			if os.IsNotExist(err) || err.Error() == "operation not supported" || err.Error() == "invalid argument" {
```
Comment (discordianfish) [inline sysfs/class_fibrechannel.go:163]: Can we add a comment why these errors are okay?

**Comment by discordianfish** [comment]
Just a last minor nit, LGTM beside that!

File: sysfs/class_fibrechannel.go (hunk 5 above)
Comment (deusnefum) [reply sysfs/class_fibrechannel.go:164]: Added!

Comment (discordianfish) [reply sysfs/class_fibrechannel.go:164]: THanks!

**Comment by deusnefum** [comment]
Is a release going to be cut soon? Seems like the easiest way for me to update node_exporter with the fibrechannel collector is to bump the procfs version used.

**Comment by discordianfish** [comment]
@deusnefum No plans for the next release yet but feel free to submit a PR to bump the procfs version.

**Pull request merged**